analytic_diffuse/models.py

```diff
@@ -38,128 +102,134 @@ def _angle_to_lmn(phi, theta):
+    {params}
     n: int
         Order of polynomial (Default of 2)
         Must be even.
+
+    {other}
     """
-    if not n % 2 == 0:
+    if n % 2:
         raise ValueError("Polynomial order must be even.")
-    return np.cos(theta) * (1 - np.sin(theta)**n)
-
-
-@checkinput
-def projgauss(phi, theta, a):
+    try:
+        return 1 - np.sin(za) ** n
+    except AttributeError:
+        return 1 - mp.sin(za) ** n
+
+@circsym
+@projected
+@sky_model
+def projgauss(az: [None, float, np.ndarray], za: [float, np.ndarray], a: float):
     """
     Gaussian weighted by cos(theta).
 
-    Parameters
-    ----------
-    phi: ndarray of float
-        Azimuth in radians, shape (Npixels,)
-    theta: ndarray of float
-        Zenith angle in radians, shape (Npixels,)
+    {params}
     a: float
         Gaussian width parameter.
+
+    {other}
+
+    Notes
+    -----
+    If called with `projected=True`, output is very similar to :func:`gauss` with
+    `projected=False`.
     """
-    return np.exp(-np.sin(theta)**2 / a**2) * np.cos(theta)
+    try:
+        return np.exp(-np.sin(za)**2 / a**2)
+    except AttributeError:
+        return mp.exp(-mp.sin(za) ** 2 / a ** 2)
 
 
-@checkinput
-def gauss(phi, theta, a, el0vec=None):
+@circsym
+@sky_model
+def gauss(az: [None, float, np.ndarray], za: [float, np.ndarray], a: float, el0vec: [None, np.ndarray]=None):
     """
     Gaussian, and off-zenith Gaussian.
 
-    Parameters
-    ----------
-    phi: ndarray of float
-        Azimuth in radians, shape (Npixels,)
-    theta: ndarray of float
-        Zenith angle in radians, shape (Npixels,)
+    {params}
     a: float
         Gaussian width parameter.
     el0vec: ndarray of float
         Cartesian vector describing lmn displacement of Gaussian center, shape (3,).
+
+    {other}
     """
+    if az is None:
+        az = np.zeros_like(za)
+
     if el0vec is None:
         el0vec = np.array([0, 0, 0])
 
     sigsq = a**2 / (2 * np.pi)
-    lmn = _angle_to_lmn(phi, theta)
-    disp = (lmn - el0vec)[:, :2]  # Only l an m
+    lmn = _angle_to_lmn(az, za)
+    disp = (lmn - el0vec)[:, :2]  # Only l and m
     ang = np.linalg.norm(disp, axis=1)
-    return np.exp(-ang**2 / (2 * sigsq))
+    try:
+        return np.exp(-ang**2 / (2 * sigsq))
+    except AttributeError:
+        return mp.exp(-ang ** 2 / (2 * sigsq))
 
 
-@checkinput
-def xysincs(phi, theta, a, xi=0.0):
+@sky_model
+def xysincs(az: [None, float, np.ndarray], za: [float, np.ndarray], a: float, xi: float=0.0):
     """
     Crossed sinc functions in l,m coordinates.
 
     Defined by:
         I(x,y) = sinc(ax) sinc(ay) cos(1 - x^2 - y^2), for |x| and |y| < 1/sqrt(2)
-               = 0, otherwise
+        = 0, otherwise
     where (x,y) is rotated from (l,m) by an angle xi.
 
-    Parameters
-    ----------
-    phi: ndarray of float
-        Azimuth in radians, shape (Npixels,)
-    theta: ndarray of float
-        Zenith angle in radians, shape (Npixels,)
+    {params}
```

Comment:
I share your concern. I'm not sure what's worse: the lowered readability, or the large amount of redundant code from repeating the same docstring. I tried to go for something that still has reasonable readability (i.e. still shows where the normal params docs goes), but it's never as nice as having it there in the source code directly. 

Yes, the `sky_model` decorator interpolates the docstrings in at import-time. So they're available in any sphinx docs we may eventually want to make, but also in `help(func)` in ipython etc. 

AFAIK there's no real way to have "inherited functions" in python so that each function inherits a base signature from some interface-defining function. The closest we can come is either this decorator syntax, or using actual classes with a `__call__` method. The problem with the latter is that you have to call it like `gauss(az, za)(a)` or `gauss()(az, za, a)` which is a bit unintuitive.

Reply:
Hm... personally, I'd rather just have the repeated docstring than sacrifice readability in the code, but it's a good point -- We're aiming for inherited behavior, so ideally we want a base function whose call signature gets modified in specific cases. I'm fine with this for now and we can think about using a callable class in the future.